rfc/rfc-95/rfc-95.md

```diff
@@ -0,0 +1,319 @@
+    if (!(splitsChange instanceof SplitsAddition)) {
+      throw new UnsupportedOperationException(
+          String.format("Unsupported split change: %s", splitsChange.getClass()));
+    }
+
+    if (splitComparator != null) {
+      List<HudiSourceSplit> newSplits = Lists.newArrayList(splitsChange.splits());
+      newSplits.sort(splitComparator);
+      splits.addAll(newSplits);
+    } else {
+      splits.addAll(splitsChange.splits());
+    }
+  }
+
+  @Override
+  public void wakeUp() {
+  }
+
+  @Override
+  public void close() throws Exception {
+    currentSplitId = null;
+    if (currentReader != null) {
+      currentReader.close();
+    }
+  }
+
+  @Override
+  public void pauseOrResumeSplits(
+      Collection<HudiSourceSplit> splitsToPause, Collection<HudiSourceSplit> splitsToResume) {
+  }
+}
+```
+
+### Hudi Source Reader
+```java 
+public class HudiSourceReader<T>
+    extends SingleThreadMultiplexSourceReaderBase<
+        RecordAndPosition<T>, T, HudiSourceSplit, HudiSourceSplit> {
+  
+  // ... 
+  @Override
+  public void start() {
+    // We request a split only if we did not get splits during the checkpoint restore.
+    // Otherwise, reader restarts will keep requesting more and more splits.
+    if (getNumberOfCurrentlyAssignedSplits() == 0) {
+      requestSplit(Collections.emptyList());
+    }
+  }
+
+  @Override
+  protected void onSplitFinished(Map<String, HudiSourceSplit> finishedSplitIds) {
+    requestSplit(Lists.newArrayList(finishedSplitIds.keySet()));
+  }
+
+  @Override
+  protected HudiSourceSplit initializedState(HudiSourceSplit split) {
+    return split;
+  }
+
+  @Override
+  protected HudiSourceSplit toSplitType(String splitId, HudiSourceSplit splitState) {
+    return splitState;
+  }
+
+  private void requestSplit(Collection<String> finishedSplitIds) {
+    context.sendSourceEventToCoordinator(new SplitRequestEvent(finishedSplitIds));
+  }
+  
+  // ...
+}
+```
+
+### Hudi SplitEnumerator
+Hudi split enumerator is responsible to monitor new hudi commit, get IncrementalInputSplits and converts to HudiSourceSplit. It functions the same as StreamReadMonitoringFunction, the difference is that
+Hudi split enumerator runs in job manager. To need the requirement of streaming ETL and back-fill data from Hudi tables  within a time range. ContinuousHudiSplitEnumerator and StaticHudiSplitEnumerator 
+will be implemented accordingly.
+
+### Hudi Split Assignment
+Hudi split assigner assign new discovered or unassigned discovered split to each of split requesters. Split will be sorted by using internal hudi commit time. Thus, splits from the same hudi commit will be processed 
+first. It will be helpful for align watermark cross each of split reader.
+
+### Hudi Watermark Emitter
+If there is an event time column stats in parquet file, we can emit watermark base on the column stats. Otherwise, we can emit watermark by using hudi commit time.
```

Comment:
A common case is generating watermark based on eventTime field, such as, 
```
create table t1(
  uuid varchar(20),
  name varchar(10),
  age int,
  `partition` varchar(20),
  ts timestamp(3),
  WATERMARK FOR ts AS ts - INTERVAL '5' SECOND
)
PARTITIONED BY (`partition`)
with (
  'connector' = 'hudi',
  ...
)
```
Anyway, if we wanna support watermark in Hudi source in this RFC, more details maybe needed, for example, supporting watermark pushdown, different watermark strategys, watermark alignment among different split reader, etc.

Reply:
Agee. Added a sub section